kernel/src/gdt.c

```diff
@@ -0,0 +1,121 @@
+#include "gdt.h"
+#include <stdint.h>
+
+// https://wiki.osdev.org/Global_Descriptor_Table
+typedef struct {
+    uint16_t limit_0_15;
+    uint16_t base_0_15;
+    uint8_t base_16_23;
+    uint8_t access;
+    uint8_t limit_16_19_and_flags;
+    uint8_t base_24_31;
+} __attribute__((packed)) GDTEntry;
+
+struct {
+    GDTEntry null;
+    GDTEntry kernel_code;
+    GDTEntry kernel_data;
+    GDTEntry null2;
+    GDTEntry user_code;
+    GDTEntry user_data;
+    GDTEntry tss_low;
+    GDTEntry tss_high;
+} __attribute__((packed)) __attribute__((aligned(4096))) g_gdt = {
+    // https://wiki.osdev.org/Global_Descriptor_Table
+    // Null segments are required
+    .null = {0, 0, 0, 0, 0, 0},
+    .null2 = {0, 0, 0, 0, 0, 0},
+    // Privilege level zero kernel segments
+    .kernel_code = {0, 0, 0, 0b10011010, 0b10100000, 0},
+    .kernel_data = {0, 0, 0, 0b10010010, 0b10100000, 0},
+    // Privilege level three user segments
+    .user_code = {0, 0, 0, 0b11111010, 0b10100000, 0},
+    .user_data = {0, 0, 0, 0b11110010, 0b10100000, 0},
+    // TSS segment
+    .tss_low = {0, 0, 0, 0b10001001, 0b10100000, 0},
+    .tss_high = {0, 0, 0, 0, 0, 0},
+};
+
+// https://wiki.osdev.org/Task_State_Segment#x86_64_Structure
+struct {
+    uint32_t reserved0;
+    // Stack pointers for different privilege levels
+    uint64_t rsp0;
+    uint64_t rsp1;
+    uint64_t rsp2;
+    uint64_t reserved1;
+    // Interrupt stack table
+    uint64_t ist1;
+    uint64_t ist2;
+    uint64_t ist3;
+    uint64_t ist4;
+    uint64_t ist5;
+    uint64_t ist6;
+    uint64_t ist7;
+    uint64_t reserved2;
+    uint16_t reserved3;
+    // IO bitmap
+    uint16_t iopb_offset;
+} __attribute__((packed)) g_tss = {0};
+
+// Setup defines in assembly
+#define STR(x) #x
+#define XSTR(s) STR(s)
+asm(".equ KERNEL_CODE_SEGMENT, " XSTR(GDT_KERNEL_CODE_SEGMENT) "\n");
+asm(".equ KERNEL_DATA_SEGMENT, " XSTR(GDT_KERNEL_DATA_SEGMENT) "\n");
+asm(".equ TSS_SEGMENT, " XSTR(GDT_TSS_SEGMENT) "\n");
+#undef XSTR
+#undef STR
+
+// Define the function set_gdt in assembly
+asm("set_gdt_and_tss:\n"
+    // Disable interrupts before setting GDT and TSS
+    "cli\n"
+    // Set pointer to GDT
+    "lgdt (%rdi)\n"
+    // Setup offset to TSS
+    "mov $TSS_SEGMENT, %ax\n"
+    "ltr %ax\n"
+    // Set all segment registers to the kernel data segment
+    "mov $KERNEL_DATA_SEGMENT, %ax\n"
+    "mov %ax, %ds\n"
+    "mov %ax, %es\n"
+    "mov %ax, %fs\n"
+    "mov %ax, %gs\n"
+    "mov %ax, %ss\n"
+    // Push code segment onto stack before return address to use far return
+    // this sets the code segment register (cs)
+    "pop %rdi\n"
```

Comment:
Unclear comments.
 
Pushing a value behind another is called inserting. 
 
`// this sets the code segment register (cs)` can either be associated with the pop or the previous comment.

Reply:
Added newlines to make it more clear